compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.analysis.getPreviousSibling
+import org.jetbrains.kotlin.fir.declarations.FirTypeParameter
+import org.jetbrains.kotlin.fir.declarations.FirTypeParameterRefsOwner
+import org.jetbrains.kotlin.fir.expressions.FirGetClassCall
+import org.jetbrains.kotlin.fir.expressions.FirResolvedQualifier
+import org.jetbrains.kotlin.fir.expressions.FirStatement
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.lexer.KtTokens.COLONCOLON
+import org.jetbrains.kotlin.lexer.KtTokens.QUEST
+
+object FirGetClassCallChecker : FirBasicExpressionChecker() {
+    override fun check(expression: FirStatement, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression !is FirGetClassCall) return
+        val source = expression.source ?: return
+        if (source.kind is FirFakeSourceElementKind) return
+
+        val argument = expression.argument as? FirResolvedQualifier ?: return
+        // Note that raw FIR drops marked nullability "?" in, e.g., A?::class, A<T?>::class, or A<T?>?::class
+        // A?::class -> CLASS_LITERAL_EXPRESSION(REFERENCE_EXPRESSION QUEST COLONCOLON "class")
+        // A<T?>::class -> ...(REFERENCE_EXPRESSION TYPE_ARGUMENT_LIST COLONCOLON "class"), and TYPE_ARGUMENT_LIST will have QUEST in it
+        // Finding `child` of QUEST can't distinguish those, though. We could filter out QUEST from TYPE_ARGUMENT_LIST, but it will miss
+        // the case like A<T?>?::class. Thus, a robust way is to the find the previous sibling of COLONCOLON (i.e., ? in the same level).
+        val beforeColonColon = source.getPreviousSibling(COLONCOLON)
+        if (argument.isNullableLHSForCallableReference || beforeColonColon?.elementType == QUEST) {
```

Comment:
Question: what is `isNullableLHSForCallableReference` and why is this check needed here?

Reply:
Class literals are a special kind of callable reference, and LHS could be an expression as well, e.g., `this::class`, `obj::prop`. Anyway, the nullability of the expression as LHS will be recorded in qualified access, during expression resolution, and we need to check that as well.